pubsublite/internal/test/verifier.go

```diff
@@ -0,0 +1,225 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package test
+
+import (
+	"container/list"
+	"sync"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/internal/testutil"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// blockWaitTimeout is the timeout for any wait operations to ensure no
+	// deadlocks.
+	blockWaitTimeout = 30 * time.Second
+)
+
+type rpcMetadata struct {
+	wantRequest   interface{}
+	retResponse   interface{}
+	retErr        error
+	blockResponse chan struct{}
+}
+
+// wait until the `blockResponse` is released by the test, or a timeout occurs.
+// Returns immediately if there was no block.
+func (r *rpcMetadata) wait() error {
+	if r.blockResponse == nil {
+		return nil
+	}
+	select {
+	case <-time.After(blockWaitTimeout):
+		return status.Errorf(codes.FailedPrecondition, "mockserver: test did not unblock response within %v", blockWaitTimeout)
+	case <-r.blockResponse:
+		return nil
+	}
+}
+
+// RPCVerifier stores an stack of requests expected from the client, and the
+// corresponding response or error to return.
+type RPCVerifier struct {
+	t        *testing.T
+	mu       sync.Mutex
+	rpcs     *list.List // Value = *rpcMetadata
+	numCalls int
+}
+
+// NewRPCVerifier creates a new verifier for requests received by the server.
+func NewRPCVerifier(t *testing.T) *RPCVerifier {
+	return &RPCVerifier{
+		t:        t,
+		rpcs:     list.New(),
+		numCalls: -1,
+	}
+}
+
+// Push appends a new {request, response, error} tuple.
+func (v *RPCVerifier) Push(wantRequest interface{}, retResponse interface{}, retErr error) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	v.rpcs.PushBack(&rpcMetadata{
+		wantRequest: wantRequest,
+		retResponse: retResponse,
+		retErr:      retErr,
+	})
+}
+
+// PushWithBlock is like Push, but returns a channel that the test should close
+// when it would like the response to be sent to the client. This is useful for
+// synchronizing with work that needs to be done on the client.
+func (v *RPCVerifier) PushWithBlock(wantRequest interface{}, retResponse interface{}, retErr error) chan struct{} {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	block := make(chan struct{})
+	v.rpcs.PushBack(&rpcMetadata{
+		wantRequest:   wantRequest,
+		retResponse:   retResponse,
+		retErr:        retErr,
+		blockResponse: block,
+	})
+	return block
+}
+
+// Pop validates the received request with the next {request, response, error}
+// tuple.
+func (v *RPCVerifier) Pop(gotRequest interface{}) (interface{}, error) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	v.numCalls++
+	elem := v.rpcs.Front()
```

Comment:
If this is a stack, shouldn't we be checking the `v.rpcs.Back()`? I think a stack would be inefficient here when implemented as a list though, so maybe it'd be better change the comments to say it's a queue instead.

Reply:
You're absolutely correct, these indeed ended up being queues. I forgot to update the comments. Fixed now.
Thanks!